### PR TITLE
Improve finitely presented semigroups

### PR DESCRIPTION
**Pull request by le27**
This PR includes the following:
* changes the input of `FpSemigroup` from a list of strings to a single string
* expands test coverage for `fpsemi.py` so that every line is checked
* fixes linting problems
* adds function to deal with brackets and powers when they are used in input of relations
* makes `FpSemigroup` a subclass of `Semigroup` and adds the `FPSOME` (Finitely Presented Semigroup Or Monoid Element) class

### CODE REVIEW

**Comment by james-d-mitchell** [comment]
The first two commits in this PR shouldn't be there since they are already in the master branch. Can you please do: 
```
git fetch origin master
git rebase origin/master
```
in your branch @le27?

File: Makefile [hunk 1]
```diff
@@ -8,7 +8,7 @@ clean:
 tests:
 	python2.7 -m nose --with-doctest
 	python3.6 -m nose --with-doctest
-
```
Comment (james-d-mitchell) [inline Makefile]: You should drop the change to this file in this commit.

File: semigroups/semifp.py [hunk 2]
```diff
@@ -43,12 +43,12 @@ def __check_word(self, word):
 
     def __init__(self, alphabet, rels):
         # Check the alphabet
-        if not isinstance(alphabet, list):
-            raise TypeError('the first argument (alphabet) must be a list')
-        elif not all(isinstance(x, str) or alphabet.count(x) > 1
+        if not isinstance(alphabet, str):
+            raise TypeError('the first argument (alphabet) must be a string')
+        elif not all(alphabet.count(x) == 1
                      for x in alphabet):
```
Comment (james-d-mitchell) [inline semigroups/semifp.py]: this should be in the previous line

File: semigroups/semifp.py [hunk 3]
```diff
@@ -119,14 +119,13 @@ def is_finite(self):
...
         # Check if any generator belongs to no relation
         for letter in self.alphabet:
+            stop = False
             for rel in self.relations:
                 for word in rel:
                     if letter in word:
                         stop = True
                         break
-                if stop:
-                    break
```
Comment (james-d-mitchell) [inline semigroups/semifp.py]: Unless I am mistaken, `break` only gets you out of the inner most loop, so by deleting these two lines you will continue looking at the next relation, when you do not have to. I think this hunk should be reverted.

File: Makefile [hunk 4]
```diff
@@ -22,7 +22,7 @@ lint:
 	flake8 semigroups/*.py 
 	flake8  tests/test_*.py
 
-doc: 
+doc:
 	cd docs ; make html; cd ..
```
Comment (james-d-mitchell) [inline Makefile]: again drop this change from this commit by reseting the file (https://stackoverflow.com/questions/215718/reset-or-revert-a-specific-file-to-a-specific-revision-using-git) and then committing and then squashing the resultant commit with this one.

File: docs/conf.py [hunk 5]
```diff
@@ -123,6 +123,8 @@
 # If true, `todo` and `todoList` produce output, else they produce nothing.
 todo_include_todos = False
 
+#Forces order of functions to appear as in source code, rather than sorted.
+autodoc_member_order = 'bysource'
```
Comment (james-d-mitchell) [inline docs/conf.py:127]: Why did you add this? It makes finding things harder in the doc. Please explain.

File: semigroups/libsemigroups.pyx [hunk 6]
```diff
@@ -477,9 +476,8 @@ cdef class SemigroupNC:
                 yield self.new_from_handle(element)
             pos += 1
 
-
 # FIXME should be a subclass of SemigroupNC
-cdef class FpSemigroupNC:
```
Comment (james-d-mitchell) [inline semigroups/libsemigroups.pyx]: Remove the FIXME, now that you did it.

File: semigroups/libsemigroups.pyx [hunk 7]
```diff
@@ -490,6 +488,7 @@ cdef class FpSemigroupNC:
         return [self.__convert_word(w) for w in rel]
 
     def __init__(self, nrgens, rels):
+
```
Comment (james-d-mitchell) [inline semigroups/libsemigroups.pyx]: Remove this blank line

File: semigroups/libsemigroups.pyx [hunk 8]
```diff
@@ -502,6 +501,7 @@ cdef class FpSemigroupNC:
         del self._rws
 
     def size(self):
+
```
Comment (james-d-mitchell) [inline semigroups/libsemigroups.pyx]: Ditto

File: semigroups/libsemigroups.pyx [hunk 9]
```diff
@@ -511,7 +511,8 @@ cdef class FpSemigroupNC:
 
     def set_report(self, val):
         '''
-        toggles whether or not to report data when running certain functions
+        Toggles whether or not to report data when running certain
```
Comment (james-d-mitchell) [inline semigroups/libsemigroups.pyx]: Toggle isn't the right word, should be `Sets` instead.

File: semigroups/libsemigroups.pyx [hunk 10]
```diff
@@ -534,14 +535,22 @@ cdef class FpSemigroupNC:
 
     def is_confluent(self):
         '''
-        check if the relations of the FpSemigroup are confluent.
+        Checks if the relations of the FpSemigroup are confluent.
```
Comment (james-d-mitchell) [inline semigroups/libsemigroups.pyx]: "Checks if the rewriting system defined by the relations of a finitely presented semigroup is confluent."

File: semigroups/libsemigroups.pyx [hunk 11]
```diff
@@ -534,14 +535,22 @@ cdef class FpSemigroupNC:
...
+        
+        If a finitely presented semigroup has a confluent rewriting system
+        then it has a solvable word problem, that is, there is an algorithm
```
Comment (james-d-mitchell) [inline semigroups/libsemigroups.pyx]: "a solvable" -> "solvable"

Comment (james-d-mitchell) [reply semigroups/libsemigroups.pyx]: Start a new sentence "...solvable word problem. In other words, ..."

File: semigroups/libsemigroups.pyx [hunk 12]
```diff
@@ -534,14 +535,22 @@ cdef class FpSemigroupNC:
 
     def is_confluent(self):
         '''
-        check if the relations of the FpSemigroup are confluent.
+        Checks if the relations of the FpSemigroup are confluent.
+        
+        If a finitely presented semigroup has a confluent rewriting system
+        then it has a solvable word problem, that is, there is an algorithm 
+        to decide when two words in the free underlying semigroup represent
```
Comment (james-d-mitchell) [inline semigroups/libsemigroups.pyx]: "...to decide when two words over the generators of the semigroup are equal."

File: semigroups/semifp.py [hunk 13]
```diff
@@ -1,9 +1,14 @@
+'''
+This module contains the classes for FpSemigroups and FpMonoids.
```
Comment (james-d-mitchell) [inline semigroups/semifp.py]: "This module contains the classes FpSemigroup and FpMonoid." (delete "for")

File: semigroups/semifp.py [hunk 14]
```diff
@@ -1,9 +1,14 @@
...
+'''
+# pylint: disable = no-member, len-as-condition
 import libsemigroups
 
-
 class FpSemigroup(libsemigroups.FpSemigroupNC):
     '''
-    A class for finitely presented semigroups.
+    A *finitely presented semigroup* is a quotient of a free semigroup on
+    a finite number of generators over a finitely generated congruence on
```
Comment (james-d-mitchell) [inline semigroups/semifp.py]: "over" -> "by", delete the words "on the free semigroup" from the end of the sentence here.

File: semigroups/semifp.py [hunk 15]
```diff
@@ -41,12 +49,139 @@ def __check_word(self, word):
...
 
+    def __remove_brackets(self, word):
+        # pylint: disable = too-many-branches
+        if word == "":
+            return ""
+
+        #if the number of left brackets is different from the number of right
+        #brackets they can't possibly pair up
+        if not word.count("(") == word.count(")"):
+            raise ValueError('invalid bracket structue')
```
Comment (james-d-mitchell) [inline semigroups/semifp.py]: Spelling, and maybe you could be more precise in the error, "unmatched bracket" or something?

File: semigroups/semifp.py [hunk 16]
```diff
@@ -41,12 +49,139 @@ def __check_word(self, word):
...
+
+        #if the ^ is at the end of the string there is no exponent.
+        #if the ^ is at the start of the string there is no base.
+        if word[0] == "^" or word[-1] == "^":
+            raise ValueError('invalid power structure')
```
Comment (james-d-mitchell) [inline semigroups/semifp.py]: haha, invalid power structure

Comment (james-d-mitchell) [reply semigroups/semifp.py]: I think we are exclusively using single quotes for strings, please replace them here.

File: semigroups/semifp.py [hunk 17]
```diff
@@ -41,12 +49,139 @@ def __check_word(self, word):
...
+        return self.__remove_powers(self.__remove_brackets(word))
+
     def __init__(self, alphabet, rels):
         # Check the alphabet
         if not isinstance(alphabet, str):
             raise TypeError('the first argument (alphabet) must be a string')
-        elif not all(alphabet.count(x) == 1
-                     for x in alphabet):
+        elif not all(alphabet.count(i) == 1
+                     for i in alphabet):
```
Comment (james-d-mitchell) [inline semigroups/semifp.py]: Won't this fit on a single line?

File: semigroups/semifp.py [hunk 18]
```diff
@@ -64,7 +199,18 @@ def __init__(self, alphabet, rels):
             raise TypeError('the second argument (relations) must be a ' +
                             'list of pairs of strings')
 
-        # TODO parse relations using ^ etc here, as per in ParseRelators in GAP
+        pure_letter_alphabet = True
+        for i in alphabet:
```
Comment (james-d-mitchell) [inline semigroups/semifp.py]: What happens if the alphabet is not pure, and the relations are not parsed, i.e. contain ^ etc?

**Review by james-d-mitchell — CHANGES_REQUESTED** [review]
I may have some more changes, but please make these ones, and I'll get back to you about the other ones.

File: semigroups/semifp.py (hunk 18 above)
Comment (le27) [reply semigroups/semifp.py]: If the alphabet is not pure then it is assumed that the relations given are already parsed as otherwise it would be ambiguous if the characters "(", ")" and "^" were being used as letters or to denote brackets or powering

Comment (james-d-mitchell) [reply semigroups/semifp.py]: Right, I guess my point was about the "assumed" part, should you not check that it is already parsed?

File: docs/conf.py (hunk 5 above)
Comment (james-d-mitchell) [reply docs/conf.py:127]: Any explanation planned?

Comment (le27) [reply docs/conf.py:127]: this was done so semigroups are listed before monoids (alex also added these lines so that his stuff would be sorted in the correct order)

**Review by james-d-mitchell — APPROVED** [review]
